make dns by querytype encoding changes configurable

derekwbrown opened the pull request:
### What does this PR do?

PR has two related changes
1) in system probe, make the list of querytypes which are recorded configurable.  Defaults to only `A` (so without adding to the configuration file, there will be no behavior change).

Configuration is as follows:
In `system_probe.yaml`
```
network_config:
  ## @param enabled - boolean - optional - default: false
  ## Set to true to enable the Network Module of the System Probe
  #
  enabled: true
  dns_recorded_query_types:
    - "A"
    - "AAAA"

```
New configuration entry is the list `dns_recorded_query_types`.  Note that must be an exact string match (case sensitive).
All entries are all upper case.
In case of misconfiguration, if no valid entries are found, code falls back to original default (`A` only).

2) in process agent, revert sending dns stats by query type
(that is, reinstates submitting by hostname, instead of by hostname/by querytype)
Adds configuration option to `process_config`: 
process_config.enable_dns_by_querytype

if set to true, new encoding is used (by hostname/by query type).

*Note* the result of this is that _both_ options therefore must be set to get any stats other than `A`.  Setting additional query types in system probe, without the process agent config, will result in the system probe _collecting_ other query types, but the information will be lost when the process agent converts to the old format _unless_ the new option is set to true.

### Motivation

Need ability to test feature prior to back-end changes.

### Additional Notes


### Describe how to test your changes

Tests should be for regression.
With a default config (neither option set), all existing DNS information for the host should be present on the app.

### Checklist

derekwbrown committed:
make dns by querytype encoding changes configurable

mbakht commented on pkg/network/dns_parser.go:
```diff
@@ -52,14 +56,15 @@ func newDNSParser(layerType gopacket.LayerType, collectDNSStats bool, collectDNS
+		dnsPayload:         dnsPayload,
+		collectDNSStats:    cfg.CollectDNSStats,
+		collectDNSDomains:  cfg.CollectDNSDomains,
+		recordedQueryTypes: queryTypes,
```
It might be helpful to log the query types being recorded.

mbakht reviewed (commented):
> Note the result of this is that both options therefore must be set to get any stats other than A

Do we need to set it up this way? We can consider supporting the case where we want to include `AAAA` or any other new query type in our stats calculation without breaking down the stats by query type.

Also, we probably should check `enable_dns_by_querytype` upstream in the `dns_parser`/`dns_stats` - otherwise, we will have a regression in terms of correctness in nodes with heavy query load if someone adds a new query type without setting `enable_dns_by_querytype` to `true`.

Let us consider the following queries:

| domain | query type |
| --- | --- |
| foo.com | AAAA |
| bar.com | AAAA|
| baz.com| AAAA|
| foo.com | A|
| bar.com|A|

Let's assume `max_dns_stats` is set to `3` and the queries are processed strictly in the above order.

Before we added support for scoping by query types, we would have reported `2` queries - only the `A` ones as `AAAA` would have been discarded in `dns_parser`.

Now, let's consider the case after this PR where someone has set `dns_recorded_query_types` to include both `A` and `AAAA` but have not set `enable_dns_by_querytype` to `true`.

`dns_stats` will process only the first `3` queries (the `AAAA` ones) and discard the rest because of `max_dns_stats=3`.
Then, in `encoder.go`, all the three stats will be discarded since none of them are `A`. We will end up reporting `0` queries where the correct answer would have been `2` (if we only support `A`) or `5` (if we support both `A` and `AAAA`).
We can get around this if we check this upstream:
- We can discard all non `A` queries in `dns_stats`/`dns_parser` when `enable_dns_by_querytype=false`. It will give us `2`
- We can set all querytype to nil/same value when `enable_dns_by_querytype=false`. It will give us `5`

derekwbrown committed:
review feedback

brycekahle commented on pkg/network/dns_parser.go:
```diff
@@ -195,7 +207,71 @@ func (*dnsParser) extractIPsInto(alias []byte, records []layers.DNSResourceRecor
+	if len(cfg.RecordedQueryTypes) > 0 {
+
+		for _, t := range cfg.RecordedQueryTypes {
+			switch t {
```
Rather than a giant switch statement, could you do something like this?
```
queryTypeStringMap := map[string]layer.DNSType{
	layers.DNSTypeA.String(): layers.DNSTypeA,
	layers.DNSTypeAAAA.String(): layers.DNSTypeAAAA,
	...
}
for _, t := range cfg.RecordedQueryTypes {
	  if qt, ok := queryTypeStringMap[t]; ok {
		  queryTypes[qt] = struct{}{}
	  } else {
		  log.Warnf("Unknown DNS type %v, skipping", t)
	  }
}
```

derekwbrown committed:
change config to more readable map dereference

brycekahle commented on pkg/network/encoding/format.go:
```diff
@@ -301,6 +308,31 @@ func formatDNSStatsByDomain(stats map[string]map[network.QueryType]network.DNSSt
+	for d, bytype := range stats {
+
+		for t, stat := range bytype {
+			if t != network.DNSTypeA {
```
I think we can remove this check. If they have configured to capture A and AAAA, but breaking stats down by query type is disabled, then we can include all stats.

derekwbrown commented:
> > Note the result of this is that both options therefore must be set to get any stats other than A
> 
> Do we need to set it up this way? We can consider supporting the case where we want to include `AAAA` or any other new query type in our stats calculation without breaking down the stats by query type.

I set it up this way because there were two distinct features that we wanted to add
- continue sending the old payloads in the current release, while allowing testing of the new payloads when the back end is ready
- wanting to be able to introduce new query types by config 

While I understand your correctness concern, that should never occur "in the wild" as there would be no point in turning on `AAAA` records in the probe if you hadn't _also_ switched to the new payload.

As I reread your message, though... it also seems reasonable to have the process config supersede the system-probe config; so if the process config is sending the old payload, then override the config and only record single-A.

brycekahle commented on pkg/config/system_probe.go:
```diff
@@ -114,6 +114,8 @@ func InitSystemProbeConfig(cfg Config) {
 	// process module
 	// nested within system_probe_config to not conflict with process-agent's process_config
 	cfg.BindEnvAndSetDefault(join(spNS, "process_config.enabled"), false, "DD_SYSTEM_PROBE_PROCESS_ENABLED")
+	cfg.SetDefault(join(netNS, "dns_recorded_query_types"), []string{})
```
Let's use `BindEnvAndSetDefault`. We want all config options to be settable from env vars. It is especially useful in containerized environments.

Can you also move this up to be with the other `network_tracer` settings?

brycekahle commented on pkg/config/config.go:
```diff
@@ -647,6 +647,7 @@ func InitConfig(config Config) {
 	}
 
 	config.BindEnv("process_config.process_dd_url", "") //nolint:errcheck
+	config.SetDefault("process_config.enable_dns_by_querytype", false)
```
I don't love that this is part of `process_config`, but since it is a temporary config option, I think it is OK.

mbakht commented:
> that should never occur "in the wild" as there would be no point in turning on AAAA records in the probe if you hadn't also switched to the new payload.

I mostly agree. However, there are two cases where we might want to turn on `AAAA` records without switching to the new payload
- If we want the total number of queries to account for both `A` and `AAAA` (similar to @brycekahle's comment above: https://github.com/DataDog/datadog-agent/pull/8586#discussion_r667146452)
- If we want to include results from `AAAA` responses to our IP->domain mapping. It does not require changes to our payload format

But I agree that it is unlikely to happen and may be not worth addressing since we will have full support for query type scoped stats by next release.

derekwbrown committed:
Fixes based on reviews
- set bydomain/byquerytype flag in system probe config (not main config)
- when multiple query types are enabled, combine them when we're not
  returning by domain/by query type
- make sure to build the domain list independent of how dns stats are
  returned

mbakht commented on pkg/network/encoding/format.go:
```diff
@@ -301,6 +308,39 @@ func formatDNSStatsByDomain(stats map[string]map[network.QueryType]network.DNSSt
+
+				m[int32(pos)] = &ms
+			}
+			break
```
Why are we doing a break here and taking only the first type if there are multiple types?

brycekahle replied:
Good catch. Yeah, this needs to be removed @derekwbrown

derekwbrown committed:
remove missed <break> from pvs. fix

mbakht commented on pkg/network/encoding/format.go:
```diff
@@ -301,6 +308,38 @@ func formatDNSStatsByDomain(stats map[string]map[network.QueryType]network.DNSSt
+
+		for _, stat := range bytype {
+
+			pos, ok := domainSet[d]
+			if !ok {
+				pos = len(domainSet)
+				domainSet[d] = pos
+			}
```
This can be moved outside the loop since this is per `domain`, not  `queryType`

olivielpeau committed:
Merge branch 'main' into db/dns-by-qtype-configurable

derekwbrown committed:
last review feedback

derekwbrown committed:
Merge branch 'db/dns-by-qtype-configurable' of github.com:DataDog/datadog-agent into db/dns-by-qtype-configurable
